/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT MODULES / SUBWORKFLOWS / FUNCTIONS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
include { FASTQC                 } from '../modules/nf-core/fastqc/main'
include { MULTIQC                } from '../modules/nf-core/multiqc/main'
include { paramsSummaryMap       } from 'plugin/nf-schema'
include { paramsSummaryMultiqc   } from '../subworkflows/nf-core/utils_nfcore_pipeline'
include { softwareVersionsToYAML } from '../subworkflows/nf-core/utils_nfcore_pipeline'
include { methodsDescriptionText } from '../subworkflows/local/utils_nfcore_precisecaller_pipeline'
include { FASTP                  } from '../modules/nf-core/fastp/main'
include { BWA_MEM                } from '../modules/nf-core/bwa/mem/main'
include { SEQKIT_SPLIT2          } from '../modules/nf-core/seqkit/split2/main'

// Create umi consensus bams from fastq
include { FASTQ_FILTER_UMI_CONSENSUS_FGBIO } from '../../subworkflows/local/fastq_filter_umi_consensus_fgbio/main'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow PRECISECALLER {
    take:
    samplesheet // channel: samplesheet read in from --input
    fasta
    fasta_fai
    dict
    bwa
    umi_file

    main:
    versions      = Channel.empty()
    multiqc_files = Channel.empty()

    /*
    ~~~~~~~~~~~~~~~~~~~~~~
       PREPROCESSING
    ~~~~~~~~~~~~~~~~~~~~~~
    */

    // Add readgroup to meta
    fastq = samplesheet.map { meta, files ->
        def rg = extractReadGroup(meta, files, params.seq_center, params.seq_platform, log)
        [ meta + [ read_group: rg ], files ]
    }

    // STEP 0: QC, UMI & TRIM //

    // MODULE: Run FastQC
    FASTQC(fastq)

    multiqc_files = multiqc_files.mix(FASTQC.out.zip.map { meta, zip -> zip })
    versions      = versions.mix(FASTQC.out.versions.first())

    // The BAM file will be generated by umi pipeline or by
    // alignment
    bam = Channel.empty()

    // Determine which preprocessing steps should be executed based on
    // user parameters. UMI processing is mutually exclusive with trimming,
    // while splitting is only applied when trimming is disabled
    enable_umi   = params.umi_read_structure
    enable_trim  = !enable_umi && !params.skip_trimming
    enable_split = !enable_trim && params.split_fastq_reads > 0

    // Alignment is required only when no BAM has been produced upstream
    enable_align = !enable_umi

    // Split FASTQ files into smaller chunks for improved parallelization
    // This step is purely structural and does not modify read sequences or
    // qualities. It is only applied when trimming is disabled
    if (enable_split) {
        SEQKIT_SPLIT2(fastq)

        fastq    = SEQKIT_SPLIT2.out.reads
        versions = versions.mix(SEQKIT_SPLIT2.out.versions)
    }

    // Perform read trimming and filtering using FastP
    // This step is skipped when UMI processing is enabled, as trimming may
    // interfere with UMI read structures
    if (enable_trim) {
        save_trimmed_fail = false
        save_merged       = false

        FASTP(
            fastq,
            [],                // no adapter
            false,             // no discard_trimmed_pass
            save_trimmed_fail,
            save_merged
        )

        fastq         = FASTP.out.reads
        versions      = versions.mix(FASTP.out.versions)
        multiqc_files = multiqc_files.mix(FASTP.out.json.map { meta, json -> json })
        multiqc_files = multiqc_files.mix(FASTP.out.html.map { meta, html -> html })
    }

    // UMI consensus calling
    if (enable_umi) {
        FASTQ_FILTER_UMI_CONSENSUS_FGBIO(
            fastq,
            fasta,
            dict,
            bwa,
            umi_file,
            params.umi_file_max_mismatches,
            params.umi_file_min_distance,
            params.group_by_umi_strategy,
            params.umi_min_reads,
            params.umi_min_base_quality,
            params.umi_max_base_error_rate
        )

        bam      = FASTQ_FILTER_UMI_CONSENSUS_FGBIO.out.bam
        versions = versions.mix(FASTQ_FILTER_UMI_CONSENSUS_FGBIO.out.versions)
    }

    // Align reads to the reference genome when UMI processing is not enabled
    // If UMI consensus calling was performed, alignment is already included
    // in the UMI subworkflow and should not be repeated
    if (enable_align) {
        sort = true
        BWA_MEM(fastq, bwa, [[id:'no_fasta'], []], sort)

        bam      = BWA_MEM.out.bam
        versions = versions.mix(BWA_MEM.out.versions)
    }






    //
    // Collate and save software versions
    //
    softwareVersionsToYAML(versions)
        .collectFile(
            storeDir: "${params.outdir}/pipeline_info",
            name:  ''  + 'pipeline_software_' +  'mqc_'  + 'versions.yml',
            sort: true,
            newLine: true
        ).set { collated_versions }

    //
    // MODULE: MultiQC
    //
    multiqc_config = Channel.fromPath("$projectDir/assets/multiqc_config.yml", checkIfExists: true)

    multiqc_custom_config = params.multiqc_config ?
        Channel.fromPath(params.multiqc_config, checkIfExists: true) :
        Channel.empty()

    multiqc_logo = params.multiqc_logo ?
        Channel.fromPath(params.multiqc_logo, checkIfExists: true) :
        Channel.empty()

    summary_params = paramsSummaryMap(workflow, parameters_schema: "nextflow_schema.json")
    workflow_summary = Channel.value(paramsSummaryMultiqc(summary_params))
    multiqc_files = multiqc_files.mix(workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))

    multiqc_custom_methods_description = params.multiqc_methods_description ?
        file(params.multiqc_methods_description, checkIfExists: true) :
        file("$projectDir/assets/methods_description_template.yml", checkIfExists: true)

    methods_description = Channel.value(methodsDescriptionText(multiqc_custom_methods_description))

    multiqc_files = multiqc_files.mix(collated_versions)
    multiqc_files = multiqc_files.mix(
        methods_description.collectFile(
            name: 'methods_description_mqc.yaml',
            sort: true
        )
    )

    MULTIQC(
        multiqc_files.collect(),
        multiqc_config.toList(),
        multiqc_custom_config.toList(),
        multiqc_logo.toList(),
        [],
        []
    )

    emit:
    multiqc_report = MULTIQC.out.report.toList() // channel: /path/to/multiqc_report.html
    versions       = versions                    // channel: [ path(versions.yml) ]
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
